src/Microsoft.AspNetCore.Blazor.Razor.Extensions/ComponentLoweringPass.cs

```diff
@@ -24,112 +24,265 @@ protected override void ExecuteCore(RazorCodeDocument codeDocument, DocumentInte
-                            };
-                        }
-                        else if (attributeNode.Children[j] is CSharpCodeAttributeValueIntermediateNode codeValue)
+                                expressionValue.Children.Single(),
+                            },
+                            Source = expressionValue.Source,
+                        };
+                    }
+                    else if (attribute.Children[i] is CSharpCodeAttributeValueIntermediateNode codeValue)
+                    {
+                        attribute.Children[i] = new CSharpExpressionIntermediateNode()
                         {
-                            attributeNode.Children[j] = new CSharpExpressionIntermediateNode()
+                            Children =
                             {
-                                Children =
-                                {
-                                    codeValue.Children.Single(),
-                                },
-                                Source = codeValue.Source,
-                            };
+                                codeValue.Children.Single(),
+                            },
+                            Source = codeValue.Source,
+                        };
+                    }
+                }
+            }
+
+            public override void VisitTagHelperProperty(TagHelperPropertyIntermediateNode node)
+            {
+                // Each 'tag helper property' belongs to a specific tag helper. We want to handle
+                // the cases for components, but leave others alone. This allows our other passes
+                // to handle those cases.
+                _children.Add(node.TagHelper.IsComponentTagHelper() ? (IntermediateNode)new ComponentAttributeExtensionNode(node) : node);
+            }
+
+            public override void VisitDefault(IntermediateNode node)
+            {
+                _children.Add(node);
+            }
+        }
+
+        private class ElementRewriteVisitor : IntermediateNodeWalker
+        {
+            private readonly IntermediateNodeCollection _children;
+
+            public ElementRewriteVisitor(IntermediateNodeCollection children)
+            {
+                _children = children;
+            }
+
+            public override void VisitTagHelper(TagHelperIntermediateNode node)
+            {
+                // Visit children, we're replacing this node.
+                for (var i = 0; i < node.Children.Count; i++)
+                {
+                    Visit(node.Children[i]);
+                }
+            }
+
+            public override void VisitTagHelperBody(TagHelperBodyIntermediateNode node)
+            {
+                for (var i = 0; i < node.Children.Count; i++)
+                {
+                    _children.Add(node.Children[i]);
+                }
+            }
+
+            public override void VisitTagHelperHtmlAttribute(TagHelperHtmlAttributeIntermediateNode node)
```

Comment:
I'm still not happy with how complicated this is, but I don't really want to spend more time on it unless there's something to be gained.